### PR TITLE
[SPARK-37371][SQL] UnionExec should support columnar if all children support columnar

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch proposes to make `UnionExec` support columnar if all its children support that.

### Why are the changes needed?

For `UnionExec`, if all its children support columnar, the union physical plan should be able to support columnar too as it is just union of all partitions from the children. By doing this, we can avoid unnecessary transition between row and columnar batch.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by viirya** [comment]
cc @cloud-fan @sunchao @dongjoon-hyun

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSetOperationsSuite.scala [hunk 1]
```diff
@@ -1355,6 +1357,45 @@ class DataFrameSetOperationsSuite extends QueryTest with SharedSparkSession {
...
+  test("SPARK-37371: UnionExec should support columnar if all children support columnar") {
+    def checkIfColumnar(
+        plan: SparkPlan,
+        targetPlan: (SparkPlan) => Boolean,
+        isColumnar: Boolean): Unit = {
+      val target = plan.collect {
+        case p if targetPlan(p) => p
+      }
+      assert(target.nonEmpty)
+      if (isColumnar) {
```
Comment (sunchao) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSetOperationsSuite.scala:1370]: nit: can replace this with one line: `assert(target.forall(_.supportsColumnar == isColumnar))`

**Review by sunchao — APPROVED** [review]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSetOperationsSuite.scala (hunk 1 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSetOperationsSuite.scala:1370]: Thanks. I'll update it.

**Comment by viirya** [comment]
Thanks. Merging to master.

**Comment by dongjoon-hyun** [comment]
+1, LGTM.

**Comment by cloud-fan** [comment]
late LGTM

**Pull request closed without merging**